Release v0.31.0 (#1884)
loadimpact/k6: release notes/v0.31.0.md

> k6 v0.31.0 is here! :tada: It's a smaller release with some performance improvements, a new `http_req_failed` metric and changes to the output subsystem that enables output extensions with xk6!

Reviewer: ```suggestion
k6 v0.31.0 is here! :tada: It's a smaller release with some performance improvements, a new `http_req_failed` metric and changes to the output subsystem that enable output extensions with [xk6](https://github.com/k6io/xk6)!
```

## New features

### Output cleanup and registration API ([#1874](https://github.com/loadimpact/k6/pull/1874))

The state of k6's output packages has been a development pain point for a long time which made it difficult to add new outputs in a consistent way. In the refactor done in v0.31.0 this has been mostly addressed and outputs now implement a saner `Output` interface.

In addition to this, an output registry was added to allow external outputs to be easily added. This expands the possibilities of what [xk6](https://github.com/k6io/xk6) can do, so now besides allowing external JS modules, it's also possible to build custom k6 binaries that use external output implementations! This is very useful if you use a system that's currently not supported by [the built-in outputs](https://k6.io/docs/getting-started/results-output), or need some custom handling of the metrics k6 produces.

> We are working on the proper documentation for this, as well as the overdue xk6 documentation about JS extensions, so keep a lookout on https://k6.io/docs/ .

Author: I considered adding a small output extension example, but it involves a lot of explanation for how e.g. the `SampleBuffer` or `PeriodicFlusher` is used, so I think it's best if we leave that for the docs.

Reviewer: A full example will be overkill here, but I added a change suggestion above with links to the new interface. I am pretty confident that for a reasonably skilled Go developer that will be enough to start with, and we can help with any questions on the forum, github or slack until we have the proper docs up...

It's now possible to declare expected HTTP response statuses for either the entire test or for individual HTTP requests, and k6 will emit a new `http_req_failed` metric as well as tag HTTP metrics with `expected_response: <bool>`.

For example:

```javascript
import http from 'k6/http';

// Set expected statuses globally for all requests.
http.setResponseCallback(http.expectedStatuses({min: 200, max: 399}, 418));

export default function () {
  // This request will be marked as failed.
  http.get('https://httpbin.test.k6.io/status/400');
  // This request will be considered as "passed" because of the responseCallback override.
  http.get('https://httpbin.test.k6.io/status/400', { responseCallback: http.expectedStatuses(400) });
}
```

Running this script will produce a summary like:

```
http_req_duration..............: avg=204.57ms min=203.31ms med=204.57ms max=205.82ms p(90)=205.57ms p(95)=205.7ms
  { expected_response:true }...: avg=203.31ms min=203.31ms med=203.31ms max=203.31ms p(90)=203.31ms p(95)=203.31ms
http_req_failed................: 50.00% ✓ 1   ✗ 1
```

Note the new `http_req_duration` sub-metric for passed requests only and the `http_req_failed` Rate metric.

> The `http.setResponseCallback()` is meant to allow arbitrary JS functions to process responses, but for now only the `http.expectedStatuses()` callback is supported.

Author: :+1: See a9e7ec3f.

## Other enhancements and UX improvements

> - JS: Because of the awesome improvements to [goja](https://github.com/dop251/goja), the JS runtime k6 uses, it's no longer necessary for k6 to load [core.js](https://github.com/zloirock/core-js) to polyfill missing JS features when using the default [`--compatibility-mode=extended`](https://k6.io/docs/using-k6/javascript-compatibility-mode). So in v0.31.0 core.js has been dropped entirely, yielding some significant CPU and memory usage improvements. The actual numbers will depend on the use case, but users can expect [TODO: Mention a percentage range?] ([#1824](https://github.com/loadimpact/k6/pull/1824))

Author: I'm hesitant to include any hard numbers in KBs, so I added some percentages in 1da8cb41 (after a quick unscientific test).

> - JS: Because of the awesome improvements to [goja](https://github.com/dop251/goja), the JS runtime k6 uses, it's no longer necessary for k6 to load [core.js](https://github.com/zloirock/core-js) to polyfill missing JS features when using the default [`--compatibility-mode=extended`](https://k6.io/docs/using-k6/javascript-compatibility-mode). So in v0.31.0 core.js has been dropped entirely, yielding some significant CPU and memory usage improvements. The actual numbers will depend on the use case, but users can expect [TODO: Mention a percentage range?] ([#1824](https://github.com/loadimpact/k6/pull/1824))

Reviewer 2: I would argue the 2MBs are more accurate as that was actually around how much parsing the whole corejs was taking so ... now that we don't do it we just don't add 2MB per VU ;) 

Also another benefit of this is that initializing VUs and consequently starting a test is now *way* faster as you don't need to execute the corejs code for each VU.

> - JS: Because of the awesome improvements to [goja](https://github.com/dop251/goja), the JS runtime k6 uses, it's no longer necessary for k6 to load [core.js](https://github.com/zloirock/core-js) to polyfill missing JS features when using the default [`--compatibility-mode=extended`](https://k6.io/docs/using-k6/javascript-compatibility-mode). So in v0.31.0 core.js has been dropped entirely, yielding some significant CPU and memory usage improvements. The actual numbers will depend on the use case, but users can expect [TODO: Mention a percentage range?] ([#1824](https://github.com/loadimpact/k6/pull/1824))

Author: Sure, see bc9c4879.

> - JS: Also because of goja improvements, some Babel plugins were disabled which should decrease CPU and memory usage even more. ([#1822](https://github.com/loadimpact/k6/pull/1822))

Reviewer 2: I wouldn't say that there is any difference really .. .the things that we disabled were mostly:
1. async/generator stuff that were just not working
2. some not very heavy plugins which likely also had little effect.
I haven't benchmarked it and it is definitely faster, just I wouldn't argue it is a something that the end user will notice :shrug:

Author: I changed it to "minor performance benefits" in 039ef774, since there should be some and we should mention it.

- JS: Expanded `ArrayBuffer` support in most internal modules, so now you can pass `ArrayBuffer` to `http.file()`, in `k6/encoding` and `k6/crypto` functions. This makes working with binary files more efficient as it doesn't require string translations. In upcoming versions we plan to expand this to the WebSocket module, as well as make some potentially breaking changes for APIs that currently return string (see the details in the Breaking Changes announcement below). ([#1800](https://github.com/loadimpact/k6/pull/1800))
- The Docker image base was updated to Alpine 3.13. Thanks @andriisoldatenko! ([#1821](https://github.com/loadimpact/k6/pull/1821))
- The Debian package now includes `ca-certificates` as a dependency. Thanks @Bablzz! ([#1854](https://github.com/loadimpact/k6/pull/1854))

## Bugs fixed!

- JS: Babel is now also run on compilation errors, which improves support of some obscure language features. ([#1861](https://github.com/loadimpact/k6/pull/1861))

> 

Reviewer: We probably should mention the concrete bugs with outputs that were fixed here - https://github.com/loadimpact/k6/issues/1423, https://github.com/loadimpact/k6/issues/1606, https://github.com/loadimpact/k6/issues/923

https://github.com/loadimpact/k6/issues/1820 should probably also be mentioned

Author: Should be resolved by afbcf9f1 and be8d9fdf.

## Internals

- JS: Some TC39 tests for unsupported features were disabled, improving the runtime of the test suite. ([#1816](https://github.com/loadimpact/k6/pull/1816))
- CI: Some more tests were enabled on Windows. ([#1855](https://github.com/loadimpact/k6/pull/1855))

## Breaking changes

> - JS: While we don't expect the core.js removal and Babel changes to impact the vast majority of users, those were substantial changes in how k6 interprets JS and a minority of users might experience issues with their tests. Please report any unexpected JavaScript errors by creating a GitHub issue.

Reviewer 2: We should probably mention that [this](https://stackoverflow.com/a/65849645/5427244) hack will no longer work due to the dropping of Promise and some babel plugins

Author: Well, it won't work as is, but they'd have to add more Babel plugins to the list?

Reviewer 2: and adding a Promise polyfill, probably, I haven't actually tried that workaround ever so :shrug: .

Reviewer 2: given that at least one person has told us they made it work and you have proposed it as a solution, I would argue we should at least say that we specifically drop some stuff that made it work and now users will need to add them back if they want to use that hack.

Author: OK, see e02862f6.

k6 v0.31.0 is here! :tada: It's a smaller release with some significant performance improvements, a new `http_req_failed` metric and changes to the output subsystem that enable output extensions with [xk6](https://github.com/k6io/xk6)!

## New features

> ### Output cleanup and registration API ([#1874](https://github.com/loadimpact/k6/pull/1874))

Reviewer: ```suggestion
### Output cleanup and extensions ([#1874](https://github.com/loadimpact/k6/pull/1874))
```

> The state of k6's output packages has been a development pain point for a long time which made it difficult to add new outputs in a consistent way. In the refactor done in v0.31.0 this has been mostly addressed and outputs now implement a saner `Output` interface.

Reviewer: ```suggestion
The state of k6's output packages has been a development pain point for a long time, which made it difficult to add new outputs in a consistent way. In the refactor done in v0.31.0, this has been mostly addressed and outputs now implement a simpler and cleaner `Output` interface.
```

> In addition to this, an output registry was added to allow external outputs to be easily added. This expands the possibilities of what [xk6](https://github.com/k6io/xk6) can do, so now besides allowing external JS modules, it's also possible to build custom k6 binaries that use external output implementations! This is very useful if you use a system that's currently not supported by [the built-in outputs](https://k6.io/docs/getting-started/results-output), or need some custom handling of the metrics k6 produces.

Reviewer: ```suggestion
In addition to this, it is now possible to implement custom k6 output extensions in Go with [xk6](https://github.com/k6io/xk6)! This is very useful if you use a system that's currently not supported by [the built-in outputs](https://k6.io/docs/getting-started/results-output), or need some custom handling of the metrics k6 produces.

Writing output extensions is done very similarly to [how JS module extensions are currently written](https://k6.io/blog/extending-k6-with-xk6), though instead of calling `js/modules.Register()`, you should implement the new [`Output` interface](https://github.com/loadimpact/k6/blob/2036baeaa83d76ce17aae95fef7e78e3ef24c735/output/types.go#L51-L74) and call [`output.RegisterExtension()`](https://github.com/loadimpact/k6/blob/2036baeaa83d76ce17aae95fef7e78e3ef24c735/output/extensions.go#L47) with your constructor.
```

### Marking requests as failed ([#1856](https://github.com/loadimpact/k6/pull/1856))

It's now possible to declare expected HTTP response statuses for either the entire test or for individual HTTP requests, and k6 will emit a new `http_req_failed` metric as well as tag HTTP metrics with `expected_response: <bool>`.

For example:

```javascript
import http from 'k6/http';

// Set expected statuses globally for all requests.
http.setResponseCallback(http.expectedStatuses({min: 200, max: 399}, 418));

export default function () {
  // This request will be marked as failed.
  http.get('https://httpbin.test.k6.io/status/400');
  // This request will be considered as "passed" because of the responseCallback override.
  http.get('https://httpbin.test.k6.io/status/400', { responseCallback: http.expectedStatuses(400) });
}
```

Running this script will produce a summary like:

```
http_req_duration..............: avg=204.57ms min=203.31ms med=204.57ms max=205.82ms p(90)=205.57ms p(95)=205.7ms
  { expected_response:true }...: avg=203.31ms min=203.31ms med=203.31ms max=203.31ms p(90)=203.31ms p(95)=203.31ms
http_req_failed................: 50.00% ✓ 1   ✗ 1
```

> Note the new `http_req_duration` sub-metric for passed requests only and the `http_req_failed` Rate metric.

Reviewer: ```suggestion
Note the new `http_req_duration` sub-metric for expected responses only, and the new `http_req_failed` `Rate` metric. This new new metric and metric tag have many potential use cases, though one of the most important ones is the ability to set better [thresholds](https://k6.io/docs/using-k6/thresholds). For example:
- `'http_req_failed': ['rate<0.1']`, i.e. fail the test if more than 10% of requests fail.
- `'http_req_duration{expected_response:true}': ['p(95)<300', 'p(99.9)<500']` - fail the test if the the 95th percentile HTTP request duration is above 300ms or the 99.9th percentile is above 500ms; specifying `expected_response:true` here may be important, because a lot of times failed requests may return more quickly than normal ones, thus skewing the results and wrongly satisfying the threshold.
```

- JS: Expanded `ArrayBuffer` support in most internal modules, so now you can pass `ArrayBuffer` to `http.file()`, in `k6/encoding` and `k6/crypto` functions. This makes working with binary files more efficient as it doesn't require string translations. In upcoming versions we plan to expand this to the WebSocket module, as well as make some potentially breaking changes for APIs that currently return an array of integers or string (see the details in the Breaking Changes announcement below). ([#1800](https://github.com/loadimpact/k6/pull/1800))
- The Docker image base was updated to Alpine 3.13. Thanks @andriisoldatenko! ([#1821](https://github.com/loadimpact/k6/pull/1821))
- The Debian package now includes `ca-certificates` as a dependency. Thanks @Bablzz! ([#1854](https://github.com/loadimpact/k6/pull/1854))

## Bugs fixed!

- Execution: Aborting a test during VU initialization (e.g. with `^C`) will now properly propagate to any used outputs. ([#1869](https://github.com/loadimpact/k6/pull/1869))

> - Execution: A race condition between the Engine and the outputs' finalization code was fixed, ensuring that all metrics are properly emitted before exiting. ([#1869](https://github.com/loadimpact/k6/pull/1869))

Reviewer: I am pretty sure this bug existed even before the `handleSummary()` changes, maybe even before the new executors :disappointed: 

```suggestion
- Execution: A race condition between the Engine and the outputs' finalization code was fixed, ensuring that all metrics are properly emitted before exiting. ([#1869](https://github.com/loadimpact/k6/pull/1869))
- Execution: Another race condition in the Engine was fixed, which may have resulted in the end-of-test summary missing some of the last test metric data. ([#1888](https://github.com/loadimpact/k6/pull/1888))
```

## Internals

- JS: `SharedArray` was rewritten using `goja.DynamicArray` making it more performant and easier to reason about. ([#1848](https://github.com/loadimpact/k6/pull/1848))
- JS: Some TC39 tests for unsupported features were disabled, improving the runtime of the test suite. ([#1816](https://github.com/loadimpact/k6/pull/1816))
- CI: Some more tests were enabled on Windows. ([#1855](https://github.com/loadimpact/k6/pull/1855))

## Breaking changes

- JS: While we don't expect the core.js removal and Babel changes to impact the vast majority of users, those were substantial changes in how k6 interprets JS and a minority of users might experience issues with their tests. Please report any unexpected JavaScript errors by creating a GitHub issue. In particular `Promise` is now `undefined`, and some unused Babel plugins like `transform-es2015-for-of` and `transform-regenerator` were also removed. This means that some workarounds like the ones mentioned [here](https://github.com/loadimpact/k6/issues/779#issuecomment-674311032) and [here](https://stackoverflow.com/a/65849645/96213) also won't work as is and will need additional polyfills and plugins to work properly.

> The following are not breaking changes in this release, but we'd like to announce them so users can prepare for upcoming releases.

Reviewer: ```suggestion
## Planned future breaking changes

The following are not breaking changes in _this_ release, but we'd like to announce them so users can prepare for them in upcoming releases (likely k6 v0.32.0).
```

> - JS: The `ArrayBuffer` changes in this release are backwards compatible and shouldn't cause any issues, but in v0.32.0 some JS APIs that currently return an array of integers or string for binary data will return `ArrayBuffer` instead. This is the case for `open()` when used with the `'b'` argument, response bodies for requests that specify `responseType: 'binary'`, `crypto.randomBytes()`, `hasher.digest('binary')`, `Response.json()` and `Response.html()` when used with requests that specify `responseType: 'binary'`, and `encoding.b64decode()`. These changes shouldn't be a problem for most users that were simply using these values to pass them to other internal modules (e.g. opening a binary file and passing it to `http.post()`), but if the scripts modified the binary data or depended on the current array of integers or string values they will need to be adapted to use [typed arrays](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Typed_arrays) instead. You can follow the discussion in [PR #1841](https://github.com/loadimpact/k6/pull/1841) and [issue #1020](https://github.com/loadimpact/k6/issues/1020).

Reviewer: ```suggestion
- JS: The `ArrayBuffer` changes in this release are backwards compatible and shouldn't cause any issues, but in v0.32.0 some JS APIs that currently return an array of integers or string for binary data will return `ArrayBuffer` instead. This is the case for `open()` when used with the `'b'` argument, response bodies for requests that specify `responseType: 'binary'`, `crypto.randomBytes()`, `hasher.digest('binary')`, and `encoding.b64decode()`. `Response.json()` and `Response.html()` will also probably stop working when used with requests that specify `responseType: 'binary'`. These changes shouldn't be a problem for most users that were simply using these values to pass them to other internal modules (e.g. opening a binary file and passing it to `http.post()`), but if the scripts modified the binary data or depended on the current array of integers or string values they will need to be adapted to use [typed arrays](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Typed_arrays) instead. You can follow the discussion in [PR #1841](https://github.com/loadimpact/k6/pull/1841) and [issue #1020](https://github.com/loadimpact/k6/issues/1020).
```

The state of k6's output packages has been a development pain point for a long time, which made it difficult to add new outputs in a consistent way. In the refactor done in v0.31.0, this has been mostly addressed and outputs now implement a simpler and cleaner `Output` interface.

In addition to this, it is now possible to implement custom k6 output extensions in Go with [xk6](https://github.com/k6io/xk6)! This is very useful if you use a system that's currently not supported by [the built-in outputs](https://k6.io/docs/getting-started/results-output), or need some custom handling of the metrics k6 produces.

Writing output extensions is done very similarly to [how JS module extensions are currently written](https://k6.io/blog/extending-k6-with-xk6), though instead of calling `js/modules.Register()`, you should implement the new [`Output` interface](https://github.com/loadimpact/k6/blob/2036baeaa83d76ce17aae95fef7e78e3ef24c735/output/types.go#L51-L74) and call [`output.RegisterExtension()`](https://github.com/loadimpact/k6/blob/2036baeaa83d76ce17aae95fef7e78e3ef24c735/output/extensions.go#L47) with your constructor.

> We are working on the proper documentation for this, as well as the overdue xk6 documentation about JS extensions, so keep a lookout on https://k6.io/docs/ .

Reviewer 3: ```suggestion
We are working on the proper documentation for this, as well as the overdue xk6 documentation about JS extensions, so keep a lookout for those on https://k6.io/docs/ .
```

> We are working on the proper documentation for this, as well as the overdue xk6 documentation about JS extensions, so keep a lookout for those on https://k6.io/docs/ .

Reviewer: ```suggestion
We are working on the proper documentation for this, as well as the overdue xk6 documentation about JS extensions, so keep a lookout for those on https://k6.io/docs/.
```

> We are working on the proper documentation for this, as well as the overdue xk6 documentation about JS extensions, so keep a lookout for those on https://k6.io/docs/ .

Reviewer: ```suggestion
We are working on the proper documentation for this, as well as the overdue xk6 documentation about JS extensions, so keep a lookout for those on [k6.io/docs/](https://k6.io/docs/).
```

### Marking requests as failed ([#1856](https://github.com/loadimpact/k6/pull/1856))

> It's now possible to declare expected HTTP response statuses for either the entire test or for individual HTTP requests, and k6 will emit a new `http_req_failed` metric as well as tag HTTP metrics with `expected_response: <bool>`.

Reviewer 3: ```suggestion
It's now possible to declare expected HTTP response statuses for either the entire test or for individual HTTP requests, and k6 will emit a new `http_req_failed` metric as well as tag HTTP metrics with `expected_response: <bool>`. By default, k6 will now fail requests that return HTTP 4xx/5xx response codes.
```

I know it's mentioned later on that the default expected is 200-399, but I think it might be worthwhile to call this out in the first paragraph as a quick summary.

For example:

```javascript
import http from 'k6/http';

// Set expected statuses globally for all requests.
http.setResponseCallback(http.expectedStatuses({min: 200, max: 399}, 418));

export default function () {
  // This request will be marked as failed.
  http.get('https://httpbin.test.k6.io/status/400');
  // This request will be considered as "passed" because of the responseCallback override.
  http.get('https://httpbin.test.k6.io/status/400', { responseCallback: http.expectedStatuses(400) });
}
```

Running this script will produce a summary like:

```
http_req_duration..............: avg=204.57ms min=203.31ms med=204.57ms max=205.82ms p(90)=205.57ms p(95)=205.7ms
  { expected_response:true }...: avg=203.31ms min=203.31ms med=203.31ms max=203.31ms p(90)=203.31ms p(95)=203.31ms
http_req_failed................: 50.00% ✓ 1   ✗ 1
```

> Note the new `http_req_duration` sub-metric for expected responses only, and the new `http_req_failed` `Rate` metric. This new new metric and metric tag have many potential use cases, though one of the most important ones is the ability to set better [thresholds](https://k6.io/docs/using-k6/thresholds). For example:

Reviewer 2: ```suggestion
Note the new `http_req_duration` sub-metric for expected responses only, and the new `http_req_failed` `Rate` metric. This new metric and metric tag have many potential use cases, though one of the most important ones is the ability to set better [thresholds](https://k6.io/docs/using-k6/thresholds). For example:
```

> Note the new `http_req_duration` sub-metric for expected responses only, and the new `http_req_failed` `Rate` metric. This new new metric and metric tag have many potential use cases, though one of the most important ones is the ability to set better [thresholds](https://k6.io/docs/using-k6/thresholds). For example:

Reviewer 3: ```suggestion
Note the new `http_req_duration` sub-metric for expected responses only, and the new `http_req_failed` `Rate` metric. This new new metric and metric tag have many potential use cases, and one of the most important ones is the ability to set better [thresholds](https://k6.io/docs/using-k6/thresholds). For example:
```

- `'http_req_failed': ['rate<0.1']`, i.e. fail the test if more than 10% of requests fail.
- `'http_req_duration{expected_response:true}': ['p(95)<300', 'p(99.9)<500']` - fail the test if the the 95th percentile HTTP request duration is above 300ms or the 99.9th percentile is above 500ms; specifying `expected_response:true` here may be important, because a lot of times failed requests may return more quickly than normal ones, thus skewing the results and wrongly satisfying the threshold.

> If the response callback is not specified, the default expected statuses will be `{min: 200, max: 399}`. The previous behavior of not emitting anything can be achieved by setting the callback to `null`, i.e. `http.setResponseCallback(null)`. Additionally the `expected_response` tag can be disabled by removing it from the default list of system tags, e.g. `k6 run --system-tags 'proto,subproto,status,method,url,name,group,check,error,error_code,tls_version,scenario,service'`.

Reviewer: ```suggestion
If the response callback is not specified, the default expected statuses will be `{min: 200, max: 399}`. The previous behavior of not emitting anything can be achieved by setting the callback to `null`, i.e. `http.setResponseCallback(null)`. Additionally, the `expected_response` tag can be disabled by removing it from the default list of [system tags](https://k6.io/docs/using-k6/options#system-tags), e.g. `k6 run --system-tags 'proto,subproto,status,method,url,name,group,check,error,error_code,tls_version,scenario,service'`.
```

> The `http.setResponseCallback()` is meant to allow arbitrary JS functions to process responses, but for now only the `http.expectedStatuses()` callback is supported.

Reviewer 2: ```suggestion
The `http.setResponseCallback()` is planned to allow arbitrary JS functions to process responses in the future, but for now only the `http.expectedStatuses()` callback is supported.
```
